File: 1966-frequency-of-the-most-frequent-element/frequency-of-the-most-frequent-element.py
class Solution:
    def maxFrequency(self, nums: list[int], k: int) -> int:
        nums.sort()
        l = 0
        total = 0
        ans = 0

        for r, val in enumerate(nums):
            total += val
            # Check if current window [l..r] is valid
            while total + k < val * (r - l + 1):
                total -= nums[l]
                l += 1
            ans = max(ans, r - l + 1)

        return ans

File: 1966-frequency-of-the-most-frequent-element/frequency-of-the-most-frequent-element.java
import java.util.Arrays;

class Solution {
    public int maxFrequency(int[] nums, int k) {
        Arrays.sort(nums);
        long total = 0;
        int ans = 0, l = 0;

        for (int r = 0; r < nums.length; r++) {
            total += nums[r];
            while (total + k < (long) nums[r] * (r - l + 1)) {
                total -= nums[l++];
            }
            ans = Math.max(ans, r - l + 1);
        }

        return ans;
    }
}
